mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md

```diff
@@ -0,0 +1,136 @@
+command line flag -affine-loop-interchange.
+
+
+### Working:
+The pass starts by locating and converting all the imperfectly nested loopnests 
+to perfectly nested ones. This process is described in next section. Once all 
+such loop nests are converted, it starts dealing with each one of them in turn:
+
+*  First, it stores all the affine.for ops in a loopnest in a llvm::SmallVector.
+It assigns loop IDs to each induction variable of the AffineForOp it encounters. 
+The loop IDs are assigned in such a way that the outermost AffineForOp induction 
+variable has an ID of 0. These IDs are helpful in calculating the access matrices 
+for load/store ops.
+
+* Next, it iterates over all the possible permutations of a loopnest to determine 
+those permutations which do not violate any dependencies in the loop nest body. 
+Such permutations are called `validPermutations`. If only one valid permutation 
+is found, we do not need to proceed any further.
+
+* In case more than one permutations are valid, we may need to consider each of 
+those permutations to detect an optimal permutation. For this, we assign a cost to
+each permutation based on a measure of locality and the extent to which the loop 
+nest can be parallelized if this permutation is followed.We select the permutation 
+with the minimum cost as the best permutation.
+
+* The cost model needs few inputs to evaluate cost of each permutation - a list  
+of affine access matrix of each load/store operation in the loopnest body. We 
+also need a vector representing which loops in the loop nest carry dependences.
+
+* Next we also need a count of the number of cache lines accessed by each 
+affine.for loop if it were considered as the innermost loop.  This information 
+is needed to get an estimate on the total number of cache lines accessed by the 
+entire loop nest for a given permutation.
+
+* Using all this information, it gets a cost for each permutation and selects 
+the one with minimum cost as the final interchange permutation. It then calls the 
+`permuteLoops` function for carrying out loop interchange.
+
+### Conversion from Imperfectly nested to Perfectly nested loopnests:
+
+For each imperfectly nested loopnest, we first construct a for-loop-tree. The 
+for-loop-tree basically captures the arrangement of for-loops in the function 
+such that all the for loops in a loop nest are grouped together in this tree.
+
+It then iterates over each loopnest in this tree and makes a separate copy of 
+each sibling (alongwith all the common parents) from rest of other siblings, 
+one at a time. From the original copy, we erase this affine.for and from the new
+copy we erase all other siblings.
+
+### Cost Model:
+The final cost of a permutation consists of a combination of three separate costs.
+These are Parallelism cost, Spatial locality cost and Temporal reuse cost.
+
+FinalCost = 100* ParallelismCost + SpatialLocalityCost + TemporalReuseCost
+
```

Comment:
How/why is the factor '100'  chosen? Document it.

Reply:
Fixed. Thanks.